design/design-generic-references.md

```diff
@@ -0,0 +1,319 @@
+# Generic Resource References
+
+* Owner: Predrag Knezevic (@pedjak)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+Crossplane providers provision and monitor external resources in third party
+cloud providers, following the declared specification in the corresponding managed
+resources. Very often, creation of these objects cannot happen in parallel, i.e. some
+fields can be set only after a given dependency is created.
+
+[Cross resource referencing] solves this problem for majority of cases by
+following [Kubernetes API object reference] convention:
+
+```yaml
+spec:
+  fooRef:
+    name: bar
+```
+
+Field `spec.fooRef` contains the name of the resource where the value for `spec.foo` is to
+be found. The kind of referred resource and the field containing the value used later for
+setting `spec.foo` is determined during compile time.
+
+Although the used convention covers the majority of use cases, it lacks the
+following properties:
+
+* The referred object kind and the source field cannot be changed at 
+  runtime for a given object
+* The referencer field may not always target a single referred kind. For
+  example, a URL field in Route53 could get populated by a URL of an S3 bucket
+  or access endpoint of an RDS Instance
+* The referencer field may not always target a single field in the referred
+  object
+* The value that is about to be assigned to the field require some sort of
+  transformation
+
+Some of the above issues can be solved by defining a composition. However, there
+are cases where one would like to use and refer to a resource not bound by a 
+composite instance.
+
+__**NOTE**__: [a prior design exists on this topic](https://github.com/crossplane/crossplane/pull/2385).
```

Comment:
I'll recap my understanding of why we want generic cross-resource references here, since it's relevant to what @muvaf wrote above. I touched on some of this at https://github.com/crossplane/crossplane/issues/1770#issuecomment-1633270957.

The current `spec.forProvider.fooRef` implementation:

1. Requires all possible references to be manually identified and declared at build time.
2. Is inflexible - users can only reference types the provider developers declared, in the ways they declared.
3. Requires the referring provider to import any types they refer to.

3 is of particular interest for provider families, because it creates a potential version dependency between two providers in a family. For example if provider-aws-eks v0.42.0 is compiled against the types from provider-aws-ec2 v0.42.0, there's a chance that installing provider-aws-ec2 v0.50.0 will break references. It would be ideal if there was no version coupling between providers within a family.